UNI-409 Improve peak detection algorithm in paramFinder

ywcui1990 opened the pull request:
Summary

* Improve peak detection algorithm in paramFinder. The paramFinder now recommends to use timeOfDay encoder for VonAnalytics data. 
* Implement a long (20 weeks) cut off on time scale for wavelet analysis
* 	Always use 'mean' aggregation unless for binary data
* Update paramFinder tests

vitaly-krugl commented on unicorn/py/unicorn_backend/param_finder.py:
```diff
@@ -44,6 +44,10 @@
 
+# Maximum time scale for wavelet analysis. The longer time scale will be ignored
+# in unit of seconds
+_MAX_TIME_SCALE = 604800 * 20
```
The constant's name should be more specific for readability. Since this is specific to wavelet analysis, the constant should probably have the word WAVELET in its name.

Is this maximum scale or maximum **window**? When I think of scale, it might be something like 1000:1 (as in maps or in drafting). If this is a window, it should have the word WINDOW in the name instead of SCALE.

If this is in units of seconds, then recommend to add the suffix `_SEC` for readability.

For example, `_MAX_WAVELET_TIME_WINDOW_SEC`

vitaly-krugl replied:
It's not obvious to the reader what period `604800` represents without doing additional math. Recommend defining and using a named constant for this:

e.g.,

`_SEVEN_DAYS_IN_SEC = 604800`

or 

`_ONE_WEEK_IN_SEC = 604800`

Then, something like 

`_MAX_WAVELET_TIME_WINDOW_SEC  = _ONE_WEEK_IN_SEC * 20`

vitaly-krugl commented:
Given that there are logic changes, shouldn't there also be test changes?

vitaly-krugl commented:
Pylint reports errors, warnings, and informational messages on this code. Need to fix them.

vitaly-krugl commented:
@ywcui1990, Please address the issues identified so far, and re-submit for code review.

vitaly-krugl commented:
Also, please use JIRA number in the PR's title to make it easier to find the JIRA from the PR. For example, this one should be renamed to "UNI-409 Improve peak detection algorithm in paramFinder"

vitaly-krugl commented on unicorn/py/unicorn_backend/param_finder.py:
```diff
@@ -488,16 +492,20 @@ def _determineEncoderTypes(cwtVar, timeScale):
+
+  # discard slow time scale (> 4 weeks ) before peak detection
+  timeScale = timeScale.astype("float32")
+  cwtVar = cwtVar[timeScale < 4*weekPeriod]
```
@ywcui1990, I am confused about what's going on in the statement `cwtVar = cwtVar[timeScale < 4*weekPeriod]` and the next similar statement as well. What is the result of `timeScale < 4*weekPeriod`? Is it just 0 or 1 or an array of values? And what does the result of `cwtVar[timeScale < 4*weekPeriod]` look like?

Perhaps some of this could also be used to improve the comments.

ywcui1990 commented:
@vitaly-krugl  
I have addressed the identified issues. Specifically,
* Updated tests for param_finder
* Improve readability of the code
* Address pylint issues

vitaly-krugl commented on unicorn/py/unicorn_backend/param_finder.py:
```diff
@@ -44,6 +42,12 @@
 _AGGREGATION_WINDOW_THRESH = 0.03
 
+# Maximum time scale for wavelet analysis. The longer time scale will be ignored
+# in unit of seconds
```
This comment applies to and needs to be next to `MAX_WAVELET_TIME_WINDOW_SEC`

vitaly-krugl commented on unicorn/py/unicorn_backend/param_finder.py:
```diff
@@ -488,19 +493,21 @@ def _determineEncoderTypes(cwtVar, timeScale):
-  cwtVarAtDayPeriod = numpy.interp(dayPeriod, timeScale, cwtVar)
-  cwtVarAtWeekPeriod = numpy.interp(weekPeriod, timeScale, cwtVar)
+  cwtVarAtDayPeriod = numpy.interp(_ONE_DAY_IN_SEC, timeScale, cwtVar)
+  cwtVarAtWeekPeriod = numpy.interp(_ONE_WEEK_IN_SEC, timeScale, cwtVar)
```
Is it okay that `_ONE_DAY_IN_SEC` and `_ONE_DAY_IN_SEC` are not floats? The old values (`weekPeriod = 604800.0` and `dayPeriod = 86400.0 `) were explicitly floating point.

vitaly-krugl commented on unicorn/py/unicorn_backend/param_finder.py:
```diff
@@ -527,46 +534,35 @@ def _determineEncoderTypes(cwtVar, timeScale):
           and cwtVarAtDayPeriod > localMaxValue * 0.5):
         useTimeOfDay = True
 
       if DISABLE_DAY_OF_WEEK_ENCODER is False:
```
`if DISABLE_DAY_OF_WEEK_ENCODER is False` ==> `if not DISABLE_DAY_OF_WEEK_ENCODER`

vitaly-krugl commented:
LGTM and okay to merge after newly noted issues are addressed.

Please note that I am unable to validate algorithmic correctness.

marionleborgne commented:
Thanks for the review @vitaly-krugl. I'll review the algorithmic correctness (@ywcui1990 and I already discussed the analysis methodology a couple of times)

marionleborgne commented:
The integration tests are passing for NAB. LGTM 👍

marionleborgne commented:
Thanks @ywcui1990

ywcui1990 commented:
I fixed the rest of issues. Thanks for the review @vitaly-krugl
